main/uart.c

```diff
@@ -15,321 +15,451 @@
-		}
-	}
-
-	//not found from the table, insert if not full
-	if (!found) {
-		if (adv_reports.num_of_advs < MAX_ADVS_TABLE) {
-			adv_reports.table[adv_reports.num_of_advs++] = *new_adv;
-		} else {
-			ret = ESP_ERR_NO_MEM;
-		}
-	}
-	portEXIT_CRITICAL(&adv_table_mux);
-
-	return ret;
+    for (size_t ii = 0; ii < binlen; ii++)
+    {
+        sprintf (hexstr + (2 * ii), "%02X", bin[ii]);
+    }
+
+    hexstr[2 * binlen] = 0;
 }
 
-int uart_send_data(const char* logName, const char* data)
+static esp_err_t adv_put_to_table (const adv_report_t * const p_adv)
 {
-	char buf[100] = { 0 };
-	const int len = strlen(data);
+    portENTER_CRITICAL (&adv_table_mux);
+    bool found = false;
+    esp_err_t ret = ESP_OK;
+
+    // Check if we already have advertisement with this MAC
+    for (int i = 0; i < adv_reports.num_of_advs; i++)
+    {
+        char * mac = adv_reports.table[i].tag_mac;
+
+        if (strcmp (p_adv->tag_mac, mac) == 0)
+        {
+            // Yes, update data.
+            found = true;
+            adv_reports.table[i] = *p_adv;
+        }
+    }
+
+    //not found from the table, insert if not full
+    if (!found)
+    {
+        if (adv_reports.num_of_advs < MAX_ADVS_TABLE)
+        {
+            adv_reports.table[adv_reports.num_of_advs++] = *p_adv;
+        }
+        else
+        {
+            ret = ESP_ERR_NO_MEM;
+        }
+    }
+
+    portEXIT_CRITICAL (&adv_table_mux);
+    return ret;
+}
 
-	for (int i = 0; i < len; i++) {
-		sprintf(buf+(2*i), "%02x", data[i]);
-	}
+int uart_send_data (const char * logName, const char * data)
+{
+    char buf[100] = { 0 };
+    const int len = strlen (data);
 
-	const int txBytes = uart_write_bytes(UART_NUM_1, data, len);
-	ESP_LOGI(logName, "Wrote to uart %d bytes, 0x%s", txBytes, buf);
+    for (int i = 0; i < len; i++)
+    {
+        sprintf (buf + (2 * i), "%02x", data[i]);
+    }
 
-	return txBytes;
+    const int txBytes = uart_write_bytes (UART_NUM_1, data, len);
+    ESP_LOGI (logName, "Wrote to uart %d bytes, 0x%s", txBytes, buf);
+    return txBytes;
 }
 
-void uart_send_nrf_command(nrf_command_t command, void* arg)
+void uart_send_nrf_command (nrf_command_t command, void * arg)
 {
-	char data[10] = { 0 };
-
-	data[0] = STX;
-
-	switch(command) {
-		case SET_FILTER: {
-			uint16_t company_id = *(uint16_t*)arg;
-			data[1] = NRF_CMD1_LEN;
-			data[2] = NRF_CMD1;
-			memcpy(&data[3], &company_id, 2);
-			data[5] = ETX;
-
-			uart_send_data(TAG, data);
-			break;
-		}
-		case CLEAR_FILTER: {
-			data[1] = NRF_CMD2_LEN;
-			data[2] = NRF_CMD2;
-			data[3] = ETX;
-
-			uart_send_data(TAG, data);
-			break;
-		}
-		default:
-		break;
-	}
+    //data[0] = STX;
+    switch (command)
+    {
+#if 0
+            char data[10] = { 0 };
+
+        case SET_FILTER:
+        {
+            uint16_t company_id = * (uint16_t *) arg;
+            data[1] = NRF_CMD1_LEN;
+            data[2] = NRF_CMD1;
+            memcpy (&data[3], &company_id, 2);
```

Comment:
this assignment is not portable, it will work differently on big-/little- endian architectures